hack/fake-apiserver/README.md

```diff
@@ -0,0 +1,265 @@
+openssl req -x509 -subj "/CN=ETCD CA" -new -newkey rsa:2048 \
+  -nodes -keyout "/tmp/etcd.key" -sha256 -days 3650 -out "/tmp/etcd.crt"
+```
+
+With the generated secrets in hands, we can request a new Kubernetes
+API endpoint as followed:
+
+```shell
+caKeyEncoded=$(cat /tmp/ca.key | base64 -w 0)
+caCertEncoded=$(cat /tmp/ca.crt | base64 -w 0)
+etcdKeyEncoded=$(cat /tmp/etcd.key | base64 -w 0)
+etcdCertEncoded=$(cat /tmp/etcd.crt | base64 -w 0)
+namespace=<cluster-namespace>
+cluster_name=<cluster-name>
+
+cluster_endpoint=$(curl -X POST "localhost:3333/register" \
+-H "Content-Type: application/json" -d '{
+  "resource": "'$namespace/$cluster_name'",
+  "caKey": "'$caKeyEncoded'",
+  "caCert": "'$caCertEncoded'",
+  "etcdKey": "'$etcdKeyEncoded'",
+  "etcdCert": "'$etcdCertEncoded'"
+}')
+```
+
+The fake API server will return a response with the ip and port of the newly
+generated api server. For example:
+
+```json
+{
+  "Host": "10.244.0.83",
+  "Port": 20000
+}
+```
+
+We also need to manually create the `ca-secret` and `etcd-secret` of the new cluster,
+so that they match the certs used by the api server.
+
+```shell
+cat <<EOF > "/tmp/${cluster}-ca-secrets.yaml"
+apiVersion: v1
+kind: Secret
+metadata:
+  labels:
+    cluster.x-k8s.io/cluster-name: ${cluster}
+  name: ${cluster}-ca
+  namespace: ${namespace}
+type: kubernetes.io/tls
+data:
+  tls.crt: ${caCertEncoded}
+  tls.key: ${caKeyEncoded}
+EOF
+
+kubectl -n ${namespace} apply -f /tmp/${cluster}-ca-secrets.yaml
+
+cat <<EOF > "/tmp/${cluster}-etcd-secrets.yaml"
+apiVersion: v1
+kind: Secret
+metadata:
+  labels:
+    cluster.x-k8s.io/cluster-name: ${cluster}
+  name: ${cluster}-etcd
+  namespace: ${namespace}
+type: kubernetes.io/tls
+data:
+  tls.crt: ${etcdCertEncoded}
+  tls.key: ${etcdKeyEncoded}
+EOF
+
+kubectl -n ${namespace} apply -f /tmp/${cluster}-etcd-secrets.yaml
+```
+
+A new cluster can be provisioned by injecting the host and port we
+got from FKAS to the cluster template provided by a CAPI infrastructure
+provider. For e.g., with CAPM3 that can be done as followed:
+
+```shell
+host=$(echo ${cluster_endpoints} | jq -r ".Host")
+port=$(echo ${cluster_endpoints} | jq -r ".Port")
+
+# Injecting the new api address into the cluster template by
+# exporting these env vars
+export CLUSTER_APIENDPOINT_HOST="${host}"
+export CLUSTER_APIENDPOINT_PORT="${port}"
+
+clusterctl generate cluster "${cluster}" \
+  --from "${CLUSTER_TEMPLATE}" \
+  --target-namespace "${namespace}" > /tmp/${cluster}-cluster.yaml
+kubectl apply -f /tmp/${cluster}-cluster.yaml
+```
+
+After the cluster is created, CAPI will expect that information like node name
+and provider ID is registered in the API server. Since our API server doesn't
+live inside the node, we will need to feed the info to it, by sending another
+POST request to `/updateNode` endpoint:
+
+```shell
+curl -X POST "localhost:3333/updateNode" -H "Content-Type: application/json" -d '{
```

Comment:
PUT for RESTFUL API when update resources

Reply:
Done